x-pack/legacy/plugins/fleet/server/routes/agents/unenroll.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import * as Joi from 'joi';
+import { FrameworkRequest } from '../../adapters/framework/adapter_types';
+import { ReturnTypeBulkUnenroll } from '../../../common/return_types';
+import { FleetServerLib } from '../../libs/types';
+
+export const createPOSTAgentsUnenrollRoute = (libs: FleetServerLib) => ({
+  method: 'POST',
+  path: '/api/fleet/agents/unenroll',
+  config: {
+    validate: {
+      payload: Joi.object({
+        ids: Joi.array()
+          .empty(false)
+          .items(Joi.string())
+          .optional(),
+        kuery: Joi.string().optional(),
+      }),
+    },
+  },
+  handler: async (
+    request: FrameworkRequest<{
+      payload: {
+        ids?: string[];
+        kuery?: string;
+      };
+    }>
+  ): Promise<ReturnTypeBulkUnenroll> => {
+    const { ids, kuery } = request.payload;
+    if ((!ids && !kuery) || (ids && kuery)) {
+      throw Boom.badRequest('You need to specify ids or kuery');
+    }
+
+    let toUnenrollIds: string[] = [];
+    if (ids) {
+      toUnenrollIds = ids;
+    }
+
+    if (kuery) {
+      let hasMore = true;
+      let page = 1;
+      while (hasMore) {
+        const response = await libs.agents.list(request.user, undefined, page++, 1000, kuery);
```

Comment:
i wonder if it makes sense to adjust the list API to allow retrieval without paging, so that this can be one call to list _all_ agents without having to page through the results?

Reply:
This could get very messy I think, many if not most production deployments will have well over the 10k returned documents limit of ES, so paging makes this easier to account for IMHO